### DISCUSSION TITLE
Why does HttpServer take a closure instead of an App instance?

### DISCUSSION THREAD

**Discussion by weezy20** [actix/actix-web#2814] [General]
```rust
#[actix_rt::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(
        || App::new()
            .service(web::resource("/").to(|| HttpResponse::Ok())))
        .bind("127.0.0.1:59090")?
        .run()
        .await
}
```
Why does `HttpServer` take a App factory and calls the very same closure on each worker thread? I am trying to understand this design, as one could very well instantiate one `App` struct and then `Clone` it across the various worker threads. Since `app_data` requires all data to be `Copy`, `Clone`, `Send`, `Sync`, I was under the impression that `App` could be cloned and sent across threads as well. 

```rust
/// The top-level builder for an Actix Web application.
pub struct App<T> {
    endpoint: T,
    services: Vec<Box<dyn AppServiceFactory>>,
    default: Option<Rc<BoxedHttpServiceFactory>>,
    factory_ref: Rc<RefCell<Option<AppRoutingFactory>>>,
    data_factories: Vec<FnDataFactory>,
    external: Vec<ResourceDef>,
    extensions: Extensions,
}
```
Is it because the `vec`s in here have something that aren't `Clone`able or `Send`able?

**Comment by fakeshadow** [comment]
actix-web uses `Send` + `Clone` factory types to construct `!Send` + `!Sync` types.
For example:
```rust
use std::rc::Rc;

use actix_web::{App, HttpServer};

fn main() {
    let rc = Rc::new(123);
    HttpServer::new(move || {
        let _ = rc.clone(); // this would not work.
        let rc = Rc::new(123); // this would.

        App::new()
    });
}
```

`app_data` also does not require any bound you just mentioned. This code works just fine:
```rust
use std::sync::Mutex;

use actix_web::{App, HttpServer};

fn main() {
    HttpServer::new(|| {
        App::new()
            .app_data(ptr())// !Send + !Sync
            .app_data(Mutex::new(())) // !Clone + !Copy
    });
}

fn ptr() -> *mut String {
    todo!()
}
```

In other word `App` is used to construct a per thread application instance. You can opt in to use thread safe smart pointers to make these application instances point to the same data.

**Reply by weezy20** [reply]
Are you sure about the `app_data` not requiring Clone/Copy? Last time I checked, `app_data(arg) ` would not accept a type that was `arg: !Clone` type and so we would have to use something like an `Arc`.

More specifically `arg` is either a `PgConnection`/`PgPool` from `sqlx`.

**Reply by fakeshadow** [reply]
Just run the code I posted and see for yourself.

**Reply by Abdillah** [reply]
> Are you sure about the app_data not requiring Clone/Copy?

It is the HttpServer closure that require `Clone`. I'm searching the discussion about how to pass from the outside non cloneable (like `Mutex` above) to the inside.
